src/sysupdate/sysupdate-transfer.h

```diff
@@ -46,6 +46,8 @@ struct Transfer {
         PartitionChange partition_change;
 };
 
+typedef int (*TransferProgress)(const Transfer *t, const Instance *inst, unsigned percentage, void *userdata);
```

Comment:
do you actually intend to pass anything other that Context as last arg?

then why not just make this specific, i.e. change prototype to:

```
typedef int (*TransferProgress)(const Context *c, const Transfer *transfer, const Instance *inst, unsigned percentage);
```

Reply:
or alternatively, maybe just add a field "context" to the Transfer object that points back to the context? that should be easier.

Reply:
The difficulty is dealing with circular dependencies. This is why I made it a `void*` initially

Unless we just forward-declare `typedef struct Context Context;` in sysupdate.h and leave all implementation details to sysupdate.c? I guess that could work, though it's not what we normally do

Reply:
yeah, forward declaration is fine.

we do things like that in pid 1 too, because "Unit" and "Manager" point to each other.